### PR TITLE
feat(observability): Vector `tap` subcommand

### PR DESCRIPTION
**Pull request by leebenson**
This PR introduces a `vector tap` CLI subcommand, communicating with the tap API added in #6610.

Closes epic #6518.

## Usage

Running:

```bash
vector tap --help
```

Yields:

```
vector-tap 0.13.0
Observe log events from topology components

USAGE:
    vector tap [OPTIONS] [components]...

FLAGS:
    -h, --help       Prints help information
    -V, --version    Prints version information

OPTIONS:
    -i, --interval <interval>    Interval to sample metrics at, in milliseconds [default: 500]
    -u, --url <url>              Vector GraphQL API server endpoint
    -l, --limit <limit>          Sample log events to the provided limit [default: 100]
    -f, --format <format>        Encoding format for logs printed to screen [default: json]  [possible values:
                                 json, yaml]

ARGS:
    <components>...    Components to observe (comma-separated; accepts glob patterns) [default: *]
```

By default, running `vector tap` will sample events for all source/transform components against a locally running Vector instance (`localhost:8686/graphql`)

`-f` can be provided as either "json" or "yaml" to control the output format. #6854 will extend this to logfmt.

Multiple, comma-separated components can be supplied as glob patterns.

## Example

To sample 10 events over a 1 sec interval against a remote endpoint, with all components prefixed with "web", run:

```bash
vector tap -l 5 -i 1000 -u <remote>/graphql "web*"
```

Which might return something like:

```json
{"message":"From web 2","timestamp":"2021-03-24T06:57:39.776877Z"}
{"message":"from web 1","timestamp":"2021-03-24T06:57:39.976529Z"}
{"message":"From web 2","timestamp":"2021-03-24T06:57:40.176341Z"}
{"message":"from web 1","timestamp":"2021-03-24T06:57:40.376128Z"}
{"message":"From web 2","timestamp":"2021-03-24T06:57:40.576249Z"}
```

## Reloading topology and unmatched components

The tap API is designed to persist over both configuration reloads, and when encountering unmatched components.

This allows for running a tap operation against a live Vector agent _before_ component(s) are connected to topology, to enable a range of debugging or observability use-cases.

The tap connection will also remain active if encountering configuration errors on a subsequent load (assuming the API was correctly configured initially). This is because the new configuration won't be applied until it's in a valid state, so the previous configuration will remain active.

## Notifications

The tap API in #6610 returns `(NOT_)MATCHED` notifications, which are intended as status messages to inform a client to update its interface accordingly if applicable.

Notifications are currently ignored in this PR, to avoid commingling with log events.

For future work, a `--verbose` option or equivalent could be added to add notifications to output. These could be printed in bold/styled to highlight them as notification messages.

Tracking in #6870.

## GraphQL subscription client bug

The tap API works, in part, by injecting a topology watch channel into GraphQL resolvers.

I discovered a bug where if a `connection_init` message isn't provided by the client prior to a `start` operation, context wouldn't be provided. This is because async-graphql's `graphql_subscription_with_data` expects the initial payload to merge into context.

Since I am using async-graphql's `Context.data_unchecked()` to pull context without checking for its existence, this resulted in a thread panic. It doesn't bring Vector down, but does show up in console logs and leaves the subscription hanging.

I think this is unexpected behaviour on the part of async-graphql, and will file a report. But it also highlighted a missing part of the initial handshake expected of GraphQL subscriptions (led, in mostpart, by Apollo's [spec in subscriptions-transport-ws](https://github.com/apollographql/subscriptions-transport-ws/blob/master/PROTOCOL.md#gql_connection_init)).

I've added an `init()` method to the internal `Subscription` handler to ensure this is sent with an empty (`{}`) payload to merge with our own internal context.

### CODE REVIEW

**Comment by leebenson** [comment]
Async-graphql bug filed + fixed: https://github.com/async-graphql/async-graphql/issues/451

Will bump per #6586 after #5872 lands.

File: lib/vector-api-client/src/gql/tap.rs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+            TapEncodingFormat::Json => Self::JSON,
+            TapEncodingFormat::Yaml => Self::YAML,
+        }
+    }
+}
+
+impl output_log_events_subscription::OutputLogEventsSubscriptionOutputLogEvents {
+    pub fn as_log(
+        &self,
+    ) -> Option<&output_log_events_subscription::OutputLogEventsSubscriptionOutputLogEventsOnLogEvent>
```
Comment (jszwedko) [inline lib/vector-api-client/src/gql/tap.rs:51]: What a type name! 🙂

File: src/top/dashboard.rs [hunk 2]
```diff
@@ -23,6 +23,15 @@ trait ThousandsFormatter {
     fn thousands_format(&self) -> String;
 }
 
+impl ThousandsFormatter for u32 {
```
Comment (jszwedko) [inline src/top/dashboard.rs:26]: Is this change related? Or did you just notice its absence? It seems like we could maybe macro-ize this over all integral types.

**Review by jszwedko — APPROVED** [review]
This looks good to me! I left a note in Slack, but copying some observations here:

* It’s occurring to me that it’d be useful to know which events come from which components when multiple are being tapped. I realize this probably needs an API change to be supported. It could be reflected here by prefixing each line with `<component_name>: `
* I know we are still only tapping the output for components. I think it might be good to design the CLI interface in such a way that it’ll be easy to add tapping events going into a component. I guess we could make out the default and then have a flag for tapping the input

File: lib/vector-api-client/src/gql/tap.rs (hunk 1 above)
Comment (leebenson) [reply lib/vector-api-client/src/gql/tap.rs:51]: Tell me about it! Took some `cargo expand`ing to figure that one out.

File: src/top/dashboard.rs (hunk 2 above)
Comment (leebenson) [reply src/top/dashboard.rs:26]: This was actually an attempt to standardise the inputs of query/subscription functions in the API client, with GraphQL resolvers in the API server.

We generally take unsigned int types (and additionally use validation) on public-facing fields to make the input interface clearer, so I thought it might be worth doing the same internally to prevent negative inputs at the type level.

This turned out to be a little redundant, since the GraphQL spec (by virtue of using JSON as its transport, I guess), has no concept of unsigned integers, being typically interpreted by a JS server as a double-precision 64-bit binary format IEEE 754, so the unsigned types are just something async-graphql offers as a serialisation guard.

So, all values sent by graphql-client are actually `i64`s; we could pass `u32` around in our top/tap client code (which this was a starting point for), but we'd still be casting as the signed variants prior to the call to the API.

Which may be completely okay, but it was an effort I felt I started in vain/out of scope, and this is a remnant of it. I think it can be pulled out for now and possibly tackled later with more enthusiastic buy-in -- possibly in the form of a macro as you suggested for this particular trait.

**Comment by leebenson** [comment]
@jszwedko:

> It’s occurring to me that it’d be useful to know which events come from which components when multiple are being tapped. I realize this probably needs an API change to be supported. It could be reflected here by prefixing each line with <component_name>:

Hmm yeah. One issue with prefixing is that the format of whatever follows is then invalid. If a user is piping JSON for parsing elsewhere, they then have to manually pull that prefix out.

I could see supporting this via an optional flag. What do you think?

> I know we are still only tapping the output for components. I think it might be good to design the CLI interface in such a way that it’ll be easy to add tapping events going into a component. I guess we could make out the default and then have a flag for tapping the input

That's exactly what I was thinking. The default is output, so a bare `vector tap` continues to work. And then when inputs are introduced, a flag can be added to toggle.

**Comment by jszwedko** [comment]
> Hmm yeah. One issue with prefixing is that the format of whatever follows is then invalid. If a user is piping JSON for parsing elsewhere, they then have to manually pull that prefix out.

That's true. Maybe we could encode it in the JSON payload itself as metadata? The disadvantage there would be that it wouldn't actually just be the event that is encoded. Maybe we could do something like:

```
{ "component_name": "foo", "component_type": "remap", "component_kind": "transform", "event": { /* actual event */ } }
```

?

This would have the added advantage of having that information easily available to machine processing if they are piping the output somewhere.

This changes the semantics of the API slightly.

I'm curious what you think. I just feel like, without an indication of where the events come from, tapping multiple components will not be very useful.

**Comment by leebenson** [comment]
> I'm curious what you think. I just feel like, without an indication of where the events come from, tapping multiple components will not be very useful.

Yeah, I hear you, and agree. I'm a little weary of mutating canonical event data, at least without some explicit opt-out to allow for returning the 'raw' logs. I think it has the potential be confusing if users aren't expecting to see the new key.

What do you think about a `--raw` option that allows the user to opt-out of post-processing, for cases where they don't need additional context/meta?

File: src/top/dashboard.rs (hunk 2 above)
Comment (jszwedko) [reply src/top/dashboard.rs:26]: Ah, gotcha, I see the other update to `top` options. I'm happy to leave this in.

**Comment by jszwedko** [comment]
> > I'm curious what you think. I just feel like, without an indication of where the events come from, tapping multiple components will not be very useful.
> 
> Yeah, I hear you, and agree. I'm a little weary of mutating canonical event data, at least without some explicit opt-out to allow for returning the 'raw' logs. I think it has the potential to confuse a parser that isn't expecting that key.
> 
> What do you think about a `--raw` option that allows the user to opt-out of post-processing, for cases where they don't need additional context/meta?

👍 I think a flag is a good idea. I'm even happy to go the other way and default to no metadata if we think the more common case is tapping a single component. Then we could add `--metadata` or something to include that as well.

**Comment by leebenson** [comment]
Noting that this PR was updated to merge in the more general `outputEvents` schema added in #6610.

> 👍 I think a flag is a good idea. I'm even happy to go the other way and default to no metadata if we think the more common case is tapping a single component. Then we could add --metadata or something to include that as well.

I like that. Since this is opt-in, I think I can get this merged and tackle as a follow-up. Tracking in #6898.

**Comment by leebenson** [comment]
All tests are passing, but the benchmarks seem stuck. I'll go ahead and merge this, since it's been queued for 1.5hrs already.
